### PR TITLE
ARROW-3893: [C++] Improve adaptive int builder performance

### PR DESCRIPTION
**Pull request by pitrou**
The strategy is two-pronged:
- scalar append to int builder appends to a small uint64_t scratch space, deferring width detection and conversion until the scratch space is full
- bulk int width detection is much improved

### CODE REVIEW

**Comment by pitrou** [comment]
Benchmark of int width detection (AMD Ryzen 7, gcc 7.3.0):
```
--------------------------------------------------------------------------------
Benchmark                                         Time           CPU Iterations
--------------------------------------------------------------------------------
BM_DetectUIntWidthNoNulls/min_time:1.000         15 us         15 us      94140    38.057GB/s
BM_DetectUIntWidthNulls/min_time:1.000           36 us         36 us      38588   15.3941GB/s
BM_DetectIntWidthNoNulls/min_time:1.000          18 us         18 us      79188    31.413GB/s
BM_DetectIntWidthNulls/min_time:1.000            41 us         41 us      34523   13.6286GB/s
```

**Comment by pitrou** [comment]
Adaptive int builder benchmarks:
* before:
```
------------------------------------------------------------------------------------------------
Benchmark                                                         Time           CPU Iterations
------------------------------------------------------------------------------------------------
BM_BuildAdaptiveIntNoNulls/repeats:2                          12244 us      12234 us         54   5.10847GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2                          12293 us      12284 us         54   5.08758GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_mean                     12269 us      12259 us         54   5.09803GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_median                   12269 us      12259 us         54   5.09803GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_stddev                      35 us         36 us         54   15.1272MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3              45940 us      45914 us         15   1.36121GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3              44678 us      44653 us         15   1.39963GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3              45632 us      45603 us         15   1.37047GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_mean         45417 us      45390 us         15    1.3771GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_median       45632 us      45603 us         15   1.37047GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_stddev         658 us        657 us         15   20.5352MB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2                         23285 us      23269 us         30   5.37198GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2                         23287 us      23272 us         30   5.37118GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_mean                    23286 us      23270 us         30   5.37158GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_median                  23286 us      23270 us         30   5.37158GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_stddev                      2 us          2 us         30   597.548kB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2             46421 us      46396 us         14   1.34707GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2             46596 us      46571 us         14     1.342GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_mean        46508 us      46483 us         14   1.34453GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_median      46508 us      46483 us         14   1.34453GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_stddev        124 us        124 us         14   3.66811MB/s
```

* after:
```
------------------------------------------------------------------------------------------------
Benchmark                                                         Time           CPU Iterations
------------------------------------------------------------------------------------------------
BM_BuildAdaptiveIntNoNulls/repeats:2                           9506 us       9495 us         74   6.58249GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2                           9527 us       9520 us         74   6.56462GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_mean                      9517 us       9508 us         74   6.57355GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_median                    9517 us       9508 us         74   6.57355GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_stddev                      15 us         18 us         74   12.9434MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3              29520 us      29499 us         24   2.11864GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3              30365 us      30345 us         24    2.0596GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3              29487 us      29470 us         24   2.12075GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_mean         29791 us      29771 us         24   2.09966GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_median       29520 us      29499 us         24   2.11864GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_stddev         498 us        497 us         24   35.5477MB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2                         18526 us      18513 us         37   6.75184GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2                         18647 us      18635 us         37   6.70782GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_mean                    18586 us      18574 us         37   6.72983GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_median                  18586 us      18574 us         37   6.72983GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_stddev                     86 us         86 us         37   31.8727MB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2             29618 us      29600 us         24   2.11141GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2             29992 us      29975 us         24   2.08503GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_mean        29805 us      29787 us         24   2.09822GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_median      29805 us      29787 us         24   2.09822GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_stddev        264 us        265 us         24   19.1034MB/s
```

**Comment by pitrou** [comment]
Dictionary builder benchmarks:
* before:
```
-----------------------------------------------------------------------------------------------
Benchmark                                                        Time           CPU Iterations
-----------------------------------------------------------------------------------------------
BM_BuildInt64DictionaryArrayRandom/repeats:2                   165 us        165 us       4178   462.219MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2                   165 us        165 us       4178   463.788MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_mean              165 us        165 us       4178   463.003MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_median            165 us        165 us       4178   463.003MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_stddev              0 us          0 us       4178   1.10959MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2               154 us        154 us       4365   495.413MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2               158 us        158 us       4365   484.158MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_mean          156 us        156 us       4365   489.786MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_median        156 us        156 us       4365   489.786MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_stddev          3 us          3 us       4365   7.95881MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2                  158 us        157 us       4451   484.478MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2                  157 us        157 us       4451   486.471MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_mean             157 us        157 us       4451   485.475MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_median           157 us        157 us       4451   485.475MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_stddev             0 us          0 us       4451   1.40936MB/s
BM_BuildStringDictionaryArray/repeats:2                        388 us        388 us       1819   257.354MB/s
BM_BuildStringDictionaryArray/repeats:2                        388 us        388 us       1819   257.319MB/s
BM_BuildStringDictionaryArray/repeats:2_mean                   388 us        388 us       1819   257.337MB/s
BM_BuildStringDictionaryArray/repeats:2_median                 388 us        388 us       1819   257.337MB/s
BM_BuildStringDictionaryArray/repeats:2_stddev                   0 us          0 us       1819   25.7869kB/s
```

* after:
```
-----------------------------------------------------------------------------------------------
Benchmark                                                        Time           CPU Iterations
-----------------------------------------------------------------------------------------------
BM_BuildInt64DictionaryArrayRandom/repeats:2                   107 us        106 us       6431   716.536MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2                   105 us        105 us       6431   725.463MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_mean              106 us        106 us       6431   720.999MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_median            106 us        106 us       6431   720.999MB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_stddev              1 us          1 us       6431   6.31234MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2                99 us         98 us       6953   775.034MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2                98 us         98 us       6953   781.114MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_mean           98 us         98 us       6953   778.074MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_median         98 us         98 us       6953   778.074MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_stddev          1 us          1 us       6953   4.29916MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2                   98 us         98 us       6837   781.993MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2                   98 us         98 us       6837   775.452MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_mean              98 us         98 us       6837   778.723MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_median            98 us         98 us       6837   778.723MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_stddev             1 us          1 us       6837   4.62541MB/s
BM_BuildStringDictionaryArray/repeats:2                        276 us        276 us       2531    362.76MB/s
BM_BuildStringDictionaryArray/repeats:2                        275 us        275 us       2531   363.518MB/s
BM_BuildStringDictionaryArray/repeats:2_mean                   276 us        275 us       2531   363.139MB/s
BM_BuildStringDictionaryArray/repeats:2_median                 276 us        275 us       2531   363.139MB/s
BM_BuildStringDictionaryArray/repeats:2_stddev                   0 us          0 us       2531   549.271kB/s
```

File: cpp/src/arrow/builder.cc [hunk 1]
```diff
@@ -499,62 +541,63 @@ Status AdaptiveUIntBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
   return Status::OK();
 }
 
-Status AdaptiveUIntBuilder::AppendValues(const uint64_t* values, int64_t length,
-                                         const uint8_t* valid_bytes) {
-  RETURN_NOT_OK(Reserve(length));
+Status AdaptiveUIntBuilder::AppendValuesInternal(const uint64_t* values, int64_t length,
+                                                 const uint8_t* valid_bytes) {
+  // See AdaptiveIntBuilder::AppendValuesInternal
+  constexpr int64_t max_chunk_size = 8192;
```
Comment (xhochy) [inline cpp/src/arrow/builder.cc]: Is this something that we should define somewhere as a global constant?

**Review by xhochy — APPROVED** [review]
+1, LGTM

Nice improvement. Not sure whether a different `max_chunk_size` might make a difference depending on the underlying CPU architecture but I guess it will only be very minor.

**Comment by pitrou** [comment]
> Not sure whether a different max_chunk_size might make a difference depending on the underlying CPU architecture but I guess it will only be very minor.

On my CPU it doesn't make much of a difference as long as it remains "reasonable" (not too small, not too large).

File: cpp/src/arrow/builder.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/builder.cc]: Yes, will do.

**Comment by wesm** [comment]
I would like to review before merging (so I can familiarize myself with this code more, since @xhochy last worked on it). I'm moving apartments this week so off-and-on the computer

**Comment by wesm** [comment]
Will review this today. Can you rebase?

**Comment by pitrou** [comment]
Rebased.

**Review by wesm — APPROVED** [review]
+1, very cool stuff, and impressive speed boost!

**Pull request closed without merging**